### DISCUSSION TITLE
Constrain maximum speed of object

### DISCUSSION THREAD

**Discussion by cswinter** [jcornaz/heron#311] [Q&A] [answered]
Is there a way to constrain the maximum speed of an object so that when acceleration is applied the velocity is clipped to that value?
It's possible to do that in a system that runs outside of the physics engine, but I don't see a way to get deterministic results when running with fixed physic timesteps of different length. With a large timestep, the physics engine might accelerate and move the object further/faster than the maximum velocity would allow for before the clipping system brings the speed down back to the maximum.

**Comment by jcornaz** [comment, marked as answer]
At the moment, there is no built-in way to do that.

What you can do is make sure to clamp the velocity at each update: 
```rust
#[derive(Component)]
struct MaxSpeed(f32);

fn enforce_max_velocity(mut bodies: Query<&mut Velocity, &MaxSpeed>) {
  for (mut velocity, MaxSpeed(max_speed)) in bodies.iter_mut() {
    velocity.linear = velocity.linear.clamp_length_max(max_speed);
  }
}
```

Of course, we could add a `MaxSpeed` component in heron.
